openssl/src/ssl/mod.rs

```diff
@@ -2944,7 +3059,68 @@ where
         }
     }
 
-    // Future work: early IO methods
+    /// Initiates the handshake.
+    ///
+    /// This will fail if `set_accept_state` or `set_connect_state` was not called first.
+    ///
+    /// This corresponds to [`SSL_do_handshake`].
+    ///
+    /// [`SSL_do_handshake`]: https://www.openssl.org/docs/manmaster/man3/SSL_do_handshake.html
+    pub fn handshake(self) -> Result<SslStream<S>, HandshakeError<S>> {
+        let mut stream = self.inner;
+        let ret = unsafe { ffi::SSL_do_handshake(stream.ssl.as_ptr()) };
+        if ret > 0 {
+            Ok(stream)
+        } else {
+            let error = stream.make_error(ret);
+            match error.code() {
+                ErrorCode::WANT_READ | ErrorCode::WANT_WRITE => {
+                    Err(HandshakeError::WouldBlock(MidHandshakeSslStream { stream, error }))
+                }
+                _ => Err(HandshakeError::Failure(MidHandshakeSslStream { stream, error })),
+            }
+        }
+    }
+
+    /// Read application data transmitted by a client before handshake
+    /// completion.
+    ///
+    /// Useful for reducing latency, but vulnerable to replay attacks. Call `set_accept_state` first.
+    ///
+    /// Returns `Ok(None)` if all early data has been read.
+    ///
+    /// This corresponds to [`SSL_read_early_data`].
+    ///
+    /// [`SSL_read_early_data`]: https://www.openssl.org/docs/manmaster/man3/SSL_read_early_data.html
+    #[cfg(ossl111)]
+    pub fn read_early_data(&mut self, buf: &mut [u8]) -> Result<Option<usize>, Error> {
```

Comment:
I'm leery of defining interfaces that are lossy wrt. the upstream API. Are we certain that this distinction is never meaningful? OpenSSL seems to have gone to some lengths to establish a convention that allows it; see also `SSL_read_ex`.

Reply:
Went ahead and did this.

Reply:
Thanks! For background, SSL_read_ex was added because SSL_read can't distinguish between 0 being returned because an error occurred and 0 being returned because you passed in a zero byte buffer. As long as we can tell the difference between error and not-error we should be fine.